### ISSUE TITLE
Copying files with recursivly with exclusions

### ISSUE THREAD

**Issue by jwoertink** [laravel-mix/laravel-mix#3171]
-   Laravel Mix Version: 6.0.39
-   Node Version (`node -v`): v16.5.0
-   NPM Version (`npm -v`): 7.19.1
-   OS: macOS 11.16 Intel

### Description:

This may be more of a question, but I'm not finding the answer in the docs... I have a directory with images that contains an uneven number of subdirectories. I want to copy the structure over to my public path keeping the directory structure as-is, and have each image added to my mix-manifest as well as ignoring specific files (in this case, dot files like `.keep`, and `.DS_Store`, etc...). Is there a way to do this?

### Steps To Reproduce:

Using an image structure like this:

```
src/
  images/
    ads/
      400x80.png
    sites/
      foo/
        logo.png
      bar/
        logo.png
      baz/
        logo.png
        specials/
          footer.gif
```

This copies over the directory structure exactly how I want...

```javascript
mix
  .copyDirectory("src/assets/images", "public/images")
```

but the mix-manifest file looks like this:

```
{
    "/images/.DS_Store": "/images/.DS_Store"
}
```

This adds paths to the manifest (and somehow ignores the dot files...), but flattens out the structure:

```javascript
mix
  .copyDirectory("src/assets/images/**", "public/images")
```

```
{
    "/images/400x80.png": "/images/400x80.png",
    "/images/logo.png": "/images/logo.png",
    "/images/footer.gif": "/images/footer.gif"
}
```

These don't match anything, and just make an empty manifest

```javascript
mix
  .copyDirectory("src/assets/images/*", "public/images")

mix
  .copyDirectory("src/assets/images/!(.*)", "public/images")
```

The end result I'm looking for is the same directory structure, but the manifest containing all of the proper paths like

```
{
    "/images/ads/400x80.png": "/images/ads/400x80.png",
    "/images/sites/foo/logo.png": "/images/sites/foo/logo.png",
    "/images/sites/bar/logo.png": "/images/sites/bar/logo.png",
    "/images/sites/baz/logo.png": "/images/sites/baz/logo.png",
    "/images/sites/baz/specials/footer.gif": "/images/sites/baz/specials/footer.gif"
}
```

** On a side note, we love using Mix, and it's been added as default in [Lucky Framework](https://luckyframework.org/guides/frontend/asset-handling). Thanks for the work that's gone in to it! ❤️

**Comment by thecrypticace** [comment]
This would probably be a good default that can be configured via a 3rd argument or something. 🤔

**Comment by jwoertink** [comment]
For sure. I could see a third arg added for ignoring some files. Then it would be just a matter of getting them in to the manifest so they can all get their cache hash, and it would be so clean! `copyDirectory("src/images", "public/images", "!(.*)")`

**Comment by thecrypticace** [comment]
I think I'd do something like this and regex match against the filename (… maybe path relative to the project root?):
```js
mix.copyDirectory("src/images", "public/images", { ignore: /^\./ })
```

or maybe an extensible function syntax (or both):

```js
mix.copyDirectory("src/images", "public/images", { ignore: file => file.isDot })
```

And we'd treat ignoring dot-files as the default.

**Comment by mrleblanc101** [comment]
Yes, I don't understand why someone would watch for change in .DS_Store files.
I understand that copying a folder instead of the individual file might copy the DS_Store, but it shouldn't be watched or added to the `mix-manifest.json`
